adapter/pkg/metrics/metrics.go

```diff
@@ -0,0 +1,215 @@
+	prometheusMetricRegistry.MustRegister(hostInfo, availableCPUs, freePhysicalMemory, usedVirtualMemory, totalVirtualMemory,
+		systemCPULoad, loadAvg, processStartTime, processOpenFDs)
+}
+
+// recordMetrics record custom golang metrics
+var recordMetrics = func(collectionInterval int32) {
+	for {
+		host, err := host.Info()
+		if handleError(err, "Failed to get host info") {
+			return
+		}
+		hostInfo.WithLabelValues(host.OS).Set(1)
+		availableCPUs.Set(float64(runtime.NumCPU()))
+
+		v, err := mem.VirtualMemory()
+		if handleError(err, "Failed to read virtual memory metrics") {
+			return
+		}
+		freePhysicalMemory.Set(float64(v.Free))
+		usedVirtualMemory.Set(float64(v.Used))
+		totalVirtualMemory.Set(float64(v.Total))
+
+		percentages, err := cpu.Percent(0, false)
+		if handleError(err, "Failed to read cpu usage metrics") || len(percentages) == 0 {
+			return
+		}
+		totalPercentage := 0.0
+		for _, p := range percentages {
+			totalPercentage += p
+		}
+		averagePercentage := totalPercentage / float64(len(percentages))
+		systemCPULoad.Set(averagePercentage)
+
+		pid := os.Getpid()
+		p, err := procfs.NewProc(pid)
+		if handleError(err, "Failed to get current process") {
+			return
+		}
+		stat, err := p.Stat()
+		if handleError(err, "Failed to get process stats") {
+			return
+		}
+		t, err := stat.StartTime()
+		if handleError(err, "Failed to read process start time") {
+			return
+		}
+		processStartTime.Set(t)
+		fds, err := p.FileDescriptorsLen()
+		if handleError(err, "Failed to read file descriptor count") {
+			return
+		}
+		processOpenFDs.Set(float64(fds))
+
+		avg, err := load.Avg()
+		if handleError(err, "Failed to read cpu load averages") {
+			return
+		}
+		loadAvg.WithLabelValues("1m").Set(avg.Load1)
+		loadAvg.WithLabelValues("5m").Set(avg.Load5)
+		loadAvg.WithLabelValues("15m").Set(avg.Load15)
+
+		// Sleep before the next measurement
+		time.Sleep(time.Duration(collectionInterval) * time.Second)
+	}
+
+}
+
+func handleError(err error, message string) bool {
+	if err != nil {
+		logger.LoggerMgw.ErrorC(logging.ErrorDetails{
+			Message:   fmt.Sprintf(message, err.Error()),
+			Severity:  logging.MINOR,
+			ErrorCode: 1109,
+		})
+		return true
+	}
+	return false
+}
+
+// StartPrometheusMetricsServer initializes and starts the metrics server to expose metrics to prometheus.
+func StartPrometheusMetricsServer(port int32, collectionInterval int32) {
+	done := make(chan struct{}) // Channel to indicate recordMetrics routine exit
+
+	// Start the Prometheus metrics server
+	go func() {
+		http.Handle("/metrics", promhttp.HandlerFor(prometheusMetricRegistry, promhttp.HandlerOpts{}))
+		err := http.ListenAndServe(":"+strconv.Itoa(int(port)), nil)
+		if err != nil {
+			logger.LoggerMgw.ErrorC(logging.ErrorDetails{
+				Message:   fmt.Sprintln("Prometheus metrics server error:", err),
+				Severity:  logging.MAJOR,
+				ErrorCode: 1110,
+			})
+		}
+	}()
+
+	for {
+		// Start the recordMetrics goroutine
+		go func() {
```

Comment:
Better to add a doc comment explaining the pattern usage of `go` to create the goroutine and the `done <- struct{}{}` and `<-done` signal for parallel(to main thread) but sequential(to each call for `recordMetrics`) execution of `recordMetrics` function.

Reply:
Fixed through PR https://github.com/wso2/product-microgateway/pull/3498